x/axelarnet/message_handler.go

```diff
@@ -0,0 +1,271 @@
+		return fmt.Errorf("unrecognized destination chain %s", destChain.Name)
+	}
+
+	if !n.IsChainActivated(ctx, destChain) {
+		return fmt.Errorf("chain %s is deactivated", destChain.Name)
+	}
+
+	if err := n.ValidateAddress(ctx, nexus.CrossChainAddress{Chain: destChain, Address: msg.DestinationAddress}); err != nil {
+		return err
+	}
+
+	// only allow sends message to EVM chains
+	if (msg.Type == nexus.TypeGeneralMessage || msg.Type == nexus.TypeGeneralMessageWithToken) &&
+		!destChain.IsFrom(evmtypes.ModuleName) {
+		return fmt.Errorf("destination chain is not an EVM chain")
+	}
+
+	switch msg.Type {
+	case nexus.TypeGeneralMessage:
+		return nil
+	case nexus.TypeGeneralMessageWithToken, nexus.TypeSendToken:
+		if !n.IsAssetRegistered(ctx, chain, token.GetDenom()) {
+			return fmt.Errorf("asset %s is not registered on chain %s", token.GetDenom(), chain.Name)
+		}
+
+		if !n.IsAssetRegistered(ctx, destChain, token.GetDenom()) {
+			return fmt.Errorf("asset %s is not registered on chain %s", token.GetDenom(), destChain.Name)
+		}
+		return nil
+	default:
+		return fmt.Errorf("unrecognized message type")
+	}
+}
+
+// OnRecvMessage handles general message from a cosmos chain
+func OnRecvMessage(ctx sdk.Context, k keeper.Keeper, ibcK keeper.IBCKeeper, n types.Nexus, b types.BankKeeper, packet ibcexported.PacketI) ibcexported.Acknowledgement {
+	// the acknowledgement is considered successful if it is a ResultAcknowledgement,
+	// follow ibc transfer convention, put byte(1) in ResultAcknowledgement to indicate success.
+	ack := channeltypes.NewResultAcknowledgement([]byte{byte(1)})
+
+	data, err := toICS20Packet(packet)
+	if err != nil {
+		return channeltypes.NewErrorAcknowledgement(err)
+	}
+
+	// skip if packet not sent to Axelar message sender account
+	if data.GetReceiver() != types.MessageSender.String() {
+		return ack
+	}
+
+	var msg Message
+	if err := json.Unmarshal([]byte(data.GetMemo()), &msg); err != nil {
+		return channeltypes.NewErrorAcknowledgement(sdkerrors.Wrapf(types.ErrGeneralMessage, "cannot unmarshal memo"))
+	}
+
+	// extract token from packet
+	token, err := extractTokenFromPacketData(ctx, ibcK, n, packet)
+
+	path := types.NewIBCPath(packet.GetDestPort(), packet.GetDestChannel())
+
+	if err := validateMessage(ctx, ibcK, n, path, msg, token); err != nil {
+		return channeltypes.NewErrorAcknowledgement(err)
+	}
+
+	sourceChain := funcs.MustOk(k.GetChainNameByIBCPath(ctx, path))
+	sourceAddress := nexus.CrossChainAddress{
+		Chain:   funcs.MustOk(n.GetChain(ctx, sourceChain)),
+		Address: data.GetSender(),
+	}
+
+	switch msg.Type {
+	case nexus.TypeGeneralMessage:
+		err = handleMessage(ctx, n, sourceAddress, msg)
+	case nexus.TypeGeneralMessageWithToken:
+		err = handleMessageWithToken(ctx, n, b, sourceAddress, msg, token)
+	case nexus.TypeSendToken:
+		err = handleTokenSent(ctx, n, b, sourceAddress, msg, token)
+	default:
+		err = sdkerrors.Wrapf(types.ErrGeneralMessage, "unrecognized Message type")
+	}
+
+	if err != nil {
```

Comment:
keep the log one layer higher to catch other errors too, such as validation issues?

Reply:
ack depress the error message, I can add debug log for validation error

Reply:
ah, sure